docs/proposals/202001_thanos_query_health_handling.md

```diff
@@ -0,0 +1,86 @@
+---
+title: Thanos Query store nodes healthiness handling
+type: proposal
+menu: proposals
+status: proposed
+owner: GiedriusS
+---
+
+### Related Tickets
+
+* https://github.com/thanos-io/thanos/issues/1651 (Response caching for Thanos)
+* https://github.com/thanos-io/thanos/pull/2072 (query: add 'sticky' nodes)
+* https://github.com/cortexproject/cortex/pull/1974 (frontend: implement cache control)
+* https://github.com/thanos-io/thanos/pull/1984 (api/v1: add cache-control header)
+
+## Summary
+
+This proposal document describes how currently the healthiness of store nodes is handled and why it should be changed to work better in terms of end response caching.
+
+It explores a few options that are available - weighs their pros and cons, and finally proposes one final variant which seems to be the best out of the possible ones.
+
+## Motivation
+
+Currently Thanos Query updates the list of healthy store nodes every 5 seconds. It does that by sending the `Info()` call via gRPC. The last successful check is noted in the `LastCheck` field. At this point if it fails then we note the error and remove it from the set of active store nodes. If that succeeds then it becomes a part of the active store set.
+
+After `--store.unhealthy-timeout` passes since `LastCheck` then it also gets removed from the UI. At this point we would forget about it completely.
+
+Every time a query is executed, we consult the active store set and send the query to all of the nodes which match according to the external labels and the min/max times that they advertise. However, if a node goes down according to our previous definition then in 5 seconds we will not send anything to it anymore. This means that we won't even be able to control this via the partial response options since no query is sent to those nodes in the first place.
+
+This is problematic in the cache of end-response caching. If we know that certain StoreAPI nodes should always be up then we always want to have not just an error (if partial response is disabled) but also the appropriate `Cache-Control` header in the response. But right now we would only get it for maximum 5 seconds after a store node would go down.
+
+Thus, this logic needs to be changed somehow. There are a few possible options:
+
+* `--store.unhealthy-timeout` could be made to apply to this case as well - we could still consider it a part of the active store set while it is still visible in the UI
```

Comment:
Let's actually number then as we referenced them as numbers afterwards.

Reply:
Converted into numbers :+1: makes sense